cmd/watch.go

```diff
@@ -62,10 +61,7 @@ func (w *watchRuns) run(event filewatcher.Event) error {
 // runSingle is similar to run. It doesn't support rerun-fails. It may be
 // possible to share runSingle with run, but the defer close on the handler
 // would require at least 3 return values, so for now it is a copy.
-func runSingle(opts *options, dir string) (*testjson.Execution, error) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
+func runSingle(ctx context.Context, cancel context.CancelFunc, opts *options, dir string) (*testjson.Execution, error) {
 	if err := opts.Validate(); err != nil {
 		return nil, err
 	}
```

Comment:
I feel like the code is wrong here

So I wouldn't have updated runSingle signature the way you do.

The idea is not not cancel the context you receive, but to cancel the context you are passing to code you call from this method.

Here is what it should be

```go
func runSingle(ctx context.Context, opts *options, dir string) (*testjson.Execution, error) { 
         ctx, cancel := context.WithCancel(ctx) 
         defer cancel() 
  
          if err := opts.Validate(); err != nil { 
                  return nil, err 
          }
// …
}
```

This implies to refactor the code you just updated.

Reply:
Mmmm maybe make sense, this way it will be canceled only this sub-context